### ISSUE TITLE
can not insert into a table with always encrypted cols using executeBatch

### ISSUE THREAD

**Issue by shayannyc25** [microsoft/mssql-jdbc#1360]
## Driver version

8.3.1.jre8-preview

## SQL Server version

Microsoft SQL Azure (RTM) - 12.0.2000.8   May 15 2020 00:47:08  

## Client Operating System

Windows 10 

## JAVA/JVM version

openjdk version "1.8.0_252"
OpenJDK Runtime Environment (Zulu 8.46.0.19-win64)-Microsoft-Azure-restricted (build 1.8.0_252-b14)
OpenJDK 64-Bit Server VM (Zulu 8.46.0.19-win64)-Microsoft-Azure-restricted (build 25.252-b14, mixed mode)

## Table schema

/****** Object:  Table [dbo].[TACCT_EventHub]    Script Date: 6/15/2020 11:10:14 PM ******/
```
SET ANSI_NULLS ON
GO

SET QUOTED_IDENTIFIER ON
GO

CREATE TABLE [dbo].[TACCT_EventHub](
	[CDE_ACCT_ENTITY] [char](15) NULL,
	[CDE_ACCT_STA] [char](1) NULL,
	[CDE_ACCT_SUB_TYP] [char](6) NULL,
	[CDE_ACCT_TYP] [char](6) NULL,
	[CDE_DATA_SRC] [char](15) NULL,
	[CDE_FIN_INST] [char](6) NULL,
	[IDN_ACCT] [char](50) COLLATE Latin1_General_BIN2 ENCRYPTED WITH (COLUMN_ENCRYPTION_KEY = [CEK1], ENCRYPTION_TYPE = Deterministic, ALGORITHM = 'AEAD_AES_256_CBC_HMAC_SHA_256') NULL,
	[IDN_ACCT_ROW] [int] NULL,
	[IDN_CREAT_PGM] [char](10) NULL,
	[IDN_CREAT_USER] [char](20) NULL,
	[IDN_DLET_USER] [char](20) NULL,
	[IDN_LST_UPDT_USER] [char](20) NULL,
	[IDN_LST_UPDT_PGM] [char](10) NULL,
	[IND_DLET] [char](1) NULL,
	[NUM_ACCT] [decimal](7, 0) ENCRYPTED WITH (COLUMN_ENCRYPTION_KEY = [CEK1], ENCRYPTION_TYPE = Deterministic, ALGORITHM = 'AEAD_AES_256_CBC_HMAC_SHA_256') NULL,
	[NUM_FA] [decimal](3, 0) ENCRYPTED WITH (COLUMN_ENCRYPTION_KEY = [CEK1], ENCRYPTION_TYPE = Deterministic, ALGORITHM = 'AEAD_AES_256_CBC_HMAC_SHA_256') NULL,
	[NUM_OFFICE] [decimal](3, 0) ENCRYPTED WITH (COLUMN_ENCRYPTION_KEY = [CEK1], ENCRYPTION_TYPE = Deterministic, ALGORITHM = 'AEAD_AES_256_CBC_HMAC_SHA_256') NULL,
	[TMS_ACCT_STA_LST_UPDT] [datetime] NULL,
	[TMS_CREAT] [datetime] NULL,
	[TMS_DLET] [datetime] NULL,
	[TMS_LST_UPDT] [datetime] NULL,
	[TMS_XTRNL_ACCT_FEFR_ID] [datetime] NULL,
	[afDPFireTS] [datetime] NULL,
	[afDCFireTS] [datetime] NULL,
	[durationBeforeSqlSave] [char](10) NULL
) ON [PRIMARY]
GO
```

## Problem description

1. Expected behaviour: trying to insert into a table with Always encrypted cols using azure java function in a batch (using SQLServerPreparedStatement.executeBatch()). Same code works fine if i change it to use auto commit and insert one row at a time  using SQLServerPreparedStatement.executeUpdate() (not batch). 

2. Actual behaviour: getting several errors 

3. Error message/stack trace:

com.microsoft.sqlserver.jdbc.SQLServerException: The incoming tabular data stream (TDS) protocol stream is incorrect. The TDS headers contained errors.

java.sql.BatchUpdateException: A result set was generated for update.

SEVERE: ConnectionID:1 ClientConnectionId: 6af635d7-434c-45eb-bb9c-3f9e3da813e3: Connection.commit: Encountered TDS_COLMETADATA (0x81). SHOWPLAN is ON, ignoring.

4. Any other details that can be helpful:

## Reproduction code

Below is a code for Java Azure Function, running latest azure function run time (3), getting events from Event Hub and inserting into a table with always encrypted cols.

```
package com.mswm.azure.java;

import com.microsoft.azure.functions.annotation.*;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.microsoft.azure.functions.*;
import java.util.*;
import java.sql.Connection;
import java.sql.Driver; 
import java.sql.DriverManager;
import java.sql.Timestamp;
import java.math.BigDecimal;
import com.microsoft.sqlserver.jdbc.SQLServerPreparedStatement;


/**
 * Azure Functions with Event Hub trigger.
 */
public class EventHubTriggerJavaToSQLDB {
    static String connectionUrl =  "jdbc:sqlserver://xxx.database.windows.net:1433;user=xxx;password=xxxxx;database=xxxx;columnEncryptionSetting=Enabled;keyVaultProviderClientId=xxxxxxxx;keyVaultProviderClientKey=xxxxx;encrypt=true;trustServerCertificate=false;sendStringParametersAsUnicode=false;hostNameInCertificate=*.database.windows.net;loginTimeout=30;DescribeInputParameters=DESCRIBEALL;DescribeOutputParameters=DESCRIBEALL";
    static Connection connection;
    static SQLServerPreparedStatement  statement;
   
    static{
        try {
                if(connection == null){
                    connection = DriverManager.getConnection(connectionUrl);
                    connection.setAutoCommit(false);
                } 
                //PreparedStatement
                if(statement == null){
                    statement = (SQLServerPreparedStatement)connection.prepareStatement("INSERT INTO [dbo].[TACCT_EventHub] values(?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)");
                }      
            }catch(Exception e){
            System.out.println("Exception while inserting="+e.toString());
           // e.printStackTrace();
        }
    }
/*@FunctionName("Warmup")
    public void warmUpFunction(ExecutionContext warmupContext){
        try{
            connection = DriverManager.getConnection(connectionUrl);
            warmupContext.getLogger().info("EventHubTriggerJavaToSQLDB.WarmUp function executed.");

        }catch (SQLException e) {
            warmupContext.getLogger().info("Exception while getting="+e.toString());
            e.printStackTrace();
        }  
    }
    */
 
    /**
     * This function will be invoked when an event is received from Event Hub.
     */
    @FunctionName("EventHubTriggerJavaToSQLDB")
    public synchronized void run(
        @EventHubTrigger(name = "message", eventHubName = "eventhub2", connection = "wmarcheventhubshayan_shayanEventHubSecPolicy_EVENTHUB", consumerGroup = "testeventhubconsumergroup", cardinality = Cardinality.MANY) String[] messages,
        final ExecutionContext context
    ){
       
        //context.getLogger().info("Java Event Hub Java to SQL DB trigger function executed.");
        //context.getLogger().info("Length:" + messages.length);
        //message.forEach(singleMessage -> context.getLogger().info(singleMessage));
        Timestamp afDCFireTS =  new Timestamp((new Date()).getTime());
                   
                try{   
                        for (int i=0; i< messages.length ;i++){
                            String rawJsonData = messages[i];
                           
                           // context.getLogger().info("Raw Data:"+rawJsonData.toString());
                            ObjectMapper oM = new ObjectMapper();
                            Map<String,Object> dataPoint = oM.readValue(rawJsonData, Map.class);
                            //context.getLogger().info("Json parsed Data:"+dataPoint.toString());

                            String CDE_ACCT_ENTITY = (String)dataPoint.get("CDE_ACCT_ENTITY");
                            String CDE_ACCT_STA = (String)dataPoint.get("CDE_ACCT_STA");
                            String CDE_ACCT_SUB_TYP =(String) dataPoint.get("CDE_ACCT_SUB_TYP");
                            String CDE_ACCT_TYP = (String)dataPoint.get("CDE_ACCT_TYP");
                            String CDE_DATA_SRC = (String)dataPoint.get("CDE_DATA_SRC");
                            String CDE_FIN_INST =(String) dataPoint.get("CDE_FIN_INST");
                            String IDN_ACCT = (String)dataPoint.get("IDN_ACCT");
                            int IDN_ACCT_ROW = Integer.parseInt((String)dataPoint.get("IDN_ACCT_ROW"));
                            String IDN_CREAT_PGM =(String) dataPoint.get("IDN_CREAT_PGM");
                            String IDN_CREAT_USER =(String) dataPoint.get("IDN_CREAT_USER");
                            String IDN_DLET_USER = (String)dataPoint.get("IDN_DLET_USER");
                            String IDN_LST_UPDT_USER = (String)dataPoint.get("IDN_LST_UPDT_USER");
                            String IDN_LST_UPDT_PGM = (String)dataPoint.get("IDN_LST_UPDT_PGM");
                            String IND_DLET = (String)dataPoint.get("IND_DLET");
                            BigDecimal NUM_ACCT = new BigDecimal((String)dataPoint.get("NUM_ACCT"));
                            NUM_ACCT = NUM_ACCT.setScale(7);
                            BigDecimal NUM_FA = new BigDecimal((String)dataPoint.get("NUM_FA"));
                            NUM_FA = NUM_FA.setScale(3);
                            BigDecimal NUM_OFFICE = new BigDecimal((String)dataPoint.get("NUM_OFFICE"));
                            NUM_OFFICE = NUM_OFFICE.setScale(3);
                            Timestamp TMS_ACCT_STA_LST_UPDT = Timestamp.valueOf((String)dataPoint.get("TMS_ACCT_STA_LST_UPDT"));
                            Timestamp TMS_CREAT =Timestamp.valueOf((String)dataPoint.get("TMS_CREAT"));
                            Timestamp TMS_DLET =Timestamp.valueOf((String)dataPoint.get("TMS_DLET"));
                            Timestamp TMS_LST_UPDT =Timestamp.valueOf((String)dataPoint.get("TMS_LST_UPDT"));
                            Timestamp TMS_XTRNL_ACCT_FEFR_ID   =Timestamp.valueOf((String)dataPoint.get("TMS_XTRNL_ACCT_FEFR_ID")); 
                            Timestamp afDPFireTS   =Timestamp.valueOf((String)dataPoint.get("afDPFireTS"));
                            
                            
                    

                            statement.setString(1,CDE_ACCT_ENTITY);
                            statement.setString(2,CDE_ACCT_STA);
                            statement.setString(3,CDE_ACCT_SUB_TYP);
                            statement.setString(4,CDE_ACCT_TYP);
                            statement.setString(5, CDE_DATA_SRC);
                            statement.setString(6,CDE_FIN_INST);
                            statement.setString(7,IDN_ACCT);
                            statement.setInt(8,IDN_ACCT_ROW);
                            statement.setString(9,IDN_CREAT_PGM);
                            statement.setString(10,IDN_CREAT_USER);
                            statement.setString(11,IDN_DLET_USER);
                            statement.setString(12,IDN_LST_UPDT_USER);
                            statement.setString(13,IDN_LST_UPDT_PGM);
                            statement.setString(14,IND_DLET);
                            statement.setBigDecimal(15,NUM_ACCT,7,0);
                            statement.setBigDecimal(16,NUM_FA,3,0);
                            statement.setBigDecimal(17,NUM_OFFICE,3,0);
                            statement.setTimestamp(18,TMS_ACCT_STA_LST_UPDT);
                            statement.setTimestamp(19,TMS_CREAT);
                            statement.setTimestamp(20,TMS_DLET);
                            statement.setTimestamp(21,TMS_LST_UPDT);
                            statement.setTimestamp(22,TMS_XTRNL_ACCT_FEFR_ID);

                            statement.setTimestamp(23,afDPFireTS);
                            statement.setTimestamp(24,afDCFireTS);
                            statement.setTimestamp(25,null);

                            statement.addBatch();
                           
                        } 
                        statement.executeBatch();
                        //int[] resultsArray = statement.executeBatch();
                        //for(int j=0;j<resultsArray.length ; j++ ){
                         //   if(resultsArray[j] != 1){

                           // }

                        //}

                        connection.commit();
                        statement.clearBatch();
                        //int  result  = statement.executeUpdate();
                            //context.getLogger().info("insert for data successfull for data row["+i+"],result code="+result);
                        context.getLogger().info("batch insert for data successfull for batch with execution time of "+afDCFireTS.toString() );
                    }
                    // Handle any errors that may have occurred.
                    catch (Exception e) {
                        context.getLogger().warning("Exception while batch inserting="+e.toString());
                        //e.printStackTrace();
                        try{
                            if(statement != null ) statement.clearBatch();
                        }catch (Exception e2){//e2.printStackTrace();
                        }
                    }  
                    /*finally{
                        try{
                         if(statement != null ) statement.close();

                         if(connection != null) connection.close();
                        }catch(Exception e2){
                            context.getLogger().warning("Exception closing statment and connection="+e2.toString());
                            statement  = null;
                            connection = null;
                        }
                    }*/
           
       }
}
```

**Comment by ulvii** [comment]
Seems like duplicate of #1301 . PR #1334 has jars attached to it, please give them a try to see if your issue is resolved.

**Comment by shayannyc25** [comment]
i tried with the jar from #1334, seems to have fixed some of the issues, but now getting these new errors:

1)
javax.crypto.AEADBadTagException
"javax.crypto.AEADBadTagException: Input too short - need tag

==============================
2) Exception while batch inserting=java.util.ConcurrentModificationException

[6/16/2020 5:39:20 PM]  at java.util.ArrayList$Itr.checkForComodification(ArrayList.java:909)
[6/16/2020 5:39:20 PM]  at java.util.ArrayList$Itr.next(ArrayList.java:859)[6/16/2020 5:39:20 PM] Function "EventHubTriggerJavaToSQLDB" (Id: 16592742-44ff-45a8-87ab-ff292fac1bd0) invoked by Java Worker
[6/16/2020 5:39:20 PM]  at com.microsoft.sqlserver.jdbc.SQLServerPreparedStatement.executeBatch(SQLServerPreparedStatement.java:2066)

[6/16/2020 5:39:20 PM]  at com.mswm.azure.java.EventHubTriggerJavaToSQLDB.run(EventHubTriggerJavaToSQLDB.java:134)
[6/16/2020 5:39:20 PM]  at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[6/16/2020 5:39:20 PM]  at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[6/16/2020 5:39:20 PM]  at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[6/16/2020 5:39:20 PM]  at java.lang.reflect.Method.invoke(Method.java:498)
[6/16/2020 5:39:20 PM]  at com.microsoft.azure.functions.worker.broker.JavaMethodInvokeInfo.invoke(JavaMethodInvokeInfo.java:22)
[6/16/2020 5:39:20 PM]  at com.microsoft.azure.functions.worker.broker.JavaMethodExecutor.execute(JavaMethodExecutor.java:54)
[6/16/2020 5:39:20 PM]  at com.microsoft.azure.functions.worker.broker.JavaFunctionBroker.invokeMethod(JavaFunctionBroker.java:53)
[6/16/2020 5:39:20 PM]  at com.microsoft.azure.functions.worker.handler.InvocationRequestHandler.execute(InvocationRequestHandler.java:33)
[6/16/2020 5:39:20 PM]  at com.microsoft.azure.functions.worker.handler.InvocationRequestHandler.execute(InvocationRequestHandler.java:10)
[6/16/2020 5:39:20 PM]  at com.microsoft.azure.functions.worker.handler.MessageHandler.handle(MessageHandler.java:45)
[6/16/2020 5:39:20 PM]  at com.microsoft.azure.functions.worker.JavaWorkerClient$StreamingMessagePeer.lambda$onNext$0(JavaWorkerClient.java:92)
[6/16/2020 5:39:20 PM]  at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
[6/16/2020 5:39:20 PM]  at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[6/16/2020 5:39:20 PM]  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[6/16/2020 5:39:20 PM]  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[6/16/2020 5:39:20 PM]  at java.lang.Thread.run(Thread.java:748)
[6/16/2020 5:39:20 PM] Executed 'Functions.EventHubTriggerJavaToSQLDB' (Succeeded, Id=16592742-44ff-45a8-87ab-ff292fac1bd0)
[6/16/2020 5:39:20 PM] Executing 'Functions.EventHubTriggerJavaToSQLDB' (Reason='', Id=b96352e7-d110-481d-8c73-8cf1810f79a7)

==============================
3)

[6/16/2020 5:39:23 PM] Exception while batch inserting=java.lang.IndexOutOfBoundsException: Index: 39, Size: 39[6/16/2020 5:39:23 PM] java.lang.IndexOutOfBoundsException: Index: 39, Size: 39
[6/16/2020 5:39:23 PM]  at java.util.ArrayList.rangeCheck(ArrayList.java:657)
[6/16/2020 5:39:23 PM]  at java.util.ArrayList.get(ArrayList.java:433)
[6/16/2020 5:39:23 PM]  at com.microsoft.sqlserver.jdbc.SQLServerPreparedStatement.doExecutePreparedStatementBatch(SQLServerPreparedStatement.java:2716)

[6/16/2020 5:39:23 PM]  at com.microsoft.sqlserver.jdbc.SQLServerPreparedStatement$PrepStmtBatchExecCmd.doExecute(SQLServerPreparedStatement.java:2676)
[6/16/2020 5:39:23 PM]  at com.microsoft.sqlserver.jdbc.TDSCommand.execute(IOBuffer.java:7349)
[6/16/2020 5:39:23 PM]  at com.microsoft.sqlserver.jdbc.SQLServerConnection.executeCommand(SQLServerConnection.java:3181)
[6/16/2020 5:39:23 PM]  at com.microsoft.sqlserver.jdbc.SQLServerStatement.executeCommand(SQLServerStatement.java:247)
[6/16/2020 5:39:23 PM]  at com.microsoft.sqlserver.jdbc.SQLServerStatement.executeStatement(SQLServerStatement.java:222)
[6/16/2020 5:39:23 PM]  at com.microsoft.sqlserver.jdbc.SQLServerPreparedStatement.executeBatch(SQLServerPreparedStatement.java:2077)
[6/16/2020 5:39:23 PM]  at com.mswm.azure.java.EventHubTriggerJavaToSQLDB.run(EventHubTriggerJavaToSQLDB.java:134)
[6/16/2020 5:39:23 PM]  at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[6/16/2020 5:39:23 PM]  at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[6/16/2020 5:39:23 PM]  at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[6/16/2020 5:39:23 PM]  at java.lang.reflect.Method.invoke(Method.java:498)
[6/16/2020 5:39:23 PM]  at com.microsoft.azure.functions.worker.broker.JavaMethodInvokeInfo.invoke(JavaMethodInvokeInfo.java:22)
[6/16/2020 5:39:23 PM]  at com.microsoft.azure.functions.worker.broker.JavaMethodExecutor.execute(JavaMethodExecutor.java:54)
[6/16/2020 5:39:23 PM]  at com.microsoft.azure.functions.worker.broker.JavaFunctionBroker.invokeMethod(JavaFunctionBroker.java:53)
[6/16/2020 5:39:23 PM] Function "EventHubTriggerJavaToSQLDB" (Id: eebdffd3-99a2-41a7-90e1-7b6d3d96cce3) invoked by Java Worker
[6/16/2020 5:39:23 PM]  at com.microsoft.azure.functions.worker.handler.InvocationRequestHandler.execute(InvocationRequestHandler.java:33)
[6/16/2020 5:39:23 PM]  at com.microsoft.azure.functions.worker.handler.InvocationRequestHandler.execute(InvocationRequestHandler.java:10)
[6/16/2020 5:39:23 PM]  at com.microsoft.azure.functions.worker.handler.MessageHandler.handle(MessageHandler.java:45)
[6/16/2020 5:39:23 PM]  at com.microsoft.azure.functions.worker.JavaWorkerClient$StreamingMessagePeer.lambda$onNext$0(JavaWorkerClient.java:92)
[6/16/2020 5:39:23 PM]  at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
[6/16/2020 5:39:23 PM]  at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[6/16/2020 5:39:23 PM]  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[6/16/2020 5:39:23 PM]  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[6/16/2020 5:39:23 PM]  at java.lang.Thread.run(Thread.java:748)
[6/16/2020 5:39:23 PM] java.lang.NullPointerException
[6/16/2020 5:39:23 PM]  at com.microsoft.sqlserver.jdbc.SQLServerPreparedStatement.doExecutePreparedStatementBatch(SQLServerPreparedStatement.java:2690)
[6/16/2020 5:39:23 PM]  at com.microsoft.sqlserver.jdbc.SQLServerPreparedStatement$PrepStmtBatchExecCmd.doExecute(SQLServerPreparedStatement.java:2676)
[6/16/2020 5:39:23 PM] Executed 'Functions.EventHubTriggerJavaToSQLDB' (Succeeded, Id=eebdffd3-99a2-41a7-90e1-7b6d3d96cce3)
[6/16/2020 5:39:23 PM]  at com.microsoft.sqlserver.jdbc.TDSCommand.execute(IOBuffer.java:7349)
[6/16/2020 5:39:23 PM]  at com.microsoft.sqlserver.jdbc.SQLServerConnection.executeCommand(SQLServerConnection.java:3181)
[6/16/2020 5:39:23 PM]  at com.microsoft.sqlserver.jdbc.SQLServerStatement.executeCommand(SQLServerStatement.java:247)
[6/16/2020 5:39:23 PM]  at com.microsoft.sqlserver.jdbc.SQLServerStatement.executeStatement(SQLServerStatement.java:222)

**Comment by shayannyc25** [comment]
i get lot of the below error:

java.lang.NullPointerException
[6/16/2020 5:39:28 PM]  at com.microsoft.sqlserver.jdbc**.SQLServerPreparedStatement.doExecutePreparedStatementBatch(SQLServerPreparedStatement.java:2690)**
[6/16/2020 5:39:28 PM]  at com.microsoft.sqlserver.jdbc.SQLServerPreparedStatement$PrepStmtBatchExecCmd.doExecute(SQLServerPreparedStatement.java:2676)
[6/16/2020 5:39:28 PM]  at com.microsoft.sqlserver.jdbc.TDSCommand.execute(IOBuffer.java:7349)
[6/16/2020 5:39:28 PM]  at com.microsoft.sqlserver.jdbc.SQLServerConnection.executeCommand(SQLServerConnection.java:3181)
[6/16/2020 5:39:28 PM]  at com.microsoft.sqlserver.jdbc.SQLServerStatement.executeCommand(SQLServerStatement.java:247)
[6/16/2020 5:39:28 PM]  at com.microsoft.sqlserver.jdbc.SQLServerStatement.executeStatement(SQLServerStatement.java:222)
[6/16/2020 5:39:28 PM]  at com.microsoft.sqlserver.jdbc.SQLServerPreparedStatement.executeBatch(SQLServerPreparedStatement.java:2077)
[6/16/2020 5:39:28 PM]  at com.mswm.azure.java.EventHubTriggerJavaToSQLDB.run(EventHubTriggerJavaToSQLDB.java:134)
[6/16/2020 5:39:28 PM]  at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[6/16/2020 5:39:28 PM]  at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[6/16/2020 5:39:28 PM]  at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[6/16/2020 5:39:28 PM]  at java.lang.reflect.Method.invoke(Method.java:498)
[6/16/2020 5:39:28 PM]  at com.microsoft.azure.functions.worker.broker.JavaMethodInvokeInfo.invoke(JavaMethodInvokeInfo.java:22)
[6/16/2020 5:39:28 PM]  at com.microsoft.azure.functions.worker.broker.JavaMethodExecutor.execute(JavaMethodExecutor.java:54)
[6/16/2020 5:39:28 PM]  at com.microsoft.azure.functions.worker.broker.JavaFunctionBroker.invokeMethod(JavaFunctionBroker.java:53)
[6/16/2020 5:39:28 PM]  at com.microsoft.azure.functions.worker.handler.InvocationRequestHandler.execute(InvocationRequestHandler.java:33)
[6/16/2020 5:39:28 PM]  at com.microsoft.azure.functions.worker.handler.InvocationRequestHandler.execute(InvocationRequestHandler.java:10)
[6/16/2020 5:39:28 PM]  at com.microsoft.azure.functions.worker.handler.MessageHandler.handle(MessageHandler.java:45)
[6/16/2020 5:39:28 PM]  at com.microsoft.azure.functions.worker.JavaWorkerClient$StreamingMessagePeer.lambda$onNext$0(JavaWorkerClient.java:92)
[6/16/2020 5:39:28 PM]  at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
[6/16/2020 5:39:28 PM]  at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[6/16/2020 5:39:28 PM]  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[6/16/2020 5:39:28 PM]  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[6/16/2020 5:39:28 PM]  at java.lang.Thread.run(Thread.java:748)

**Comment by ulvii** [comment]
I see. To speed up the investigations, can you create a standalone JDBC repro code, it is almost impossible to debug on Azure  functions.

**Comment by shayannyc25** [comment]
i have given all the code for the function in this ticket, you can just copy and past the body of the function into a public static main(..) java method to make it stand alone app and test , thanks lot

**Comment by ulvii** [comment]
Where does `messages` come from? Can you  also post the data? 

I also noticed you have `SHOWPLAN is ON`, I don't think the driver is designed to work when this feature is on, especially Always Encrypted, because it needs to execute a query to fetch crypto metadata, try turning SHOWPLAN OFF.

**Comment by shayannyc25** [comment]
messages is an string array that eventhub passes to the azure java funciton, and each item in the array is one event in the event hub. We are using the default event hub batching size of ten, so the array size would be ten. It looks like as there are lots messages in the event hub (more than 10), multiple instance of the function are created and run in parallel.  
The data in the event is in json. Attached is a sample. i dont set the showplan , i dont know who is doing it.
[event hub pe data.txt](https://github.com/microsoft/mssql-jdbc/files/4794194/event.hub.pe.data.txt)

**Comment by ulvii** [comment]
Please see [this page](https://docs.microsoft.com/en-us/sql/t-sql/statements/set-showplan-all-transact-sql?view=sql-server-ver15) for details about  SHOWPLAN. It is set on the database. You can execute `SET SHOWPLAN_ALL OFF` to turn it off.

**Comment by shayannyc25** [comment]
is that set to on by default?

**Comment by ulvii** [comment]
No, it should be off by default. 

So I spent some time converting your repro code to plain JDBC code (only one message) and it works perfectly fine for me. 

```
public class Issue1360 {
    public static void main(String[] args) throws SQLException {
        Timestamp afDCFireTS = new Timestamp((new Date()).getTime());
        String url = "jdbc:sqlserver://localhost;username=user;password=ppp;database=tempdb;columnEncryptionSetting=Enabled;keyVaultProviderClientId=xxxx;keyVaultProviderClientKey=xxxx;sendStringParametersAsUnicode=false;";
        Connection connection = DriverManager.getConnection(url);

        SQLServerPreparedStatement statement = (SQLServerPreparedStatement) connection.prepareStatement(
                "INSERT INTO [dbo].[TACCT_EventHub] values(?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)");
        Map<String, Object> dataPoint = new HashMap<String, Object>();

        dataPoint.put("CDE_DATA_SRC", "tttttttttttttt4");
        dataPoint.put("IDN_LST_UPDT_PGM", "tttttttt11");
        dataPoint.put("afDCFireTS", null);
        dataPoint.put("TMS_CREAT", "2020-06-17 15:06:01.29");
        dataPoint.put("IDN_CREAT_USER", "ttttttttttttttttttt8");
        dataPoint.put("TMS_DLET", "2020-06-17 15:06:01.29");
        dataPoint.put("TMS_LST_UPDT", "2020-06-17 15:06:01.29");
        dataPoint.put("IDN_ACCT_ROW", "1");
        dataPoint.put("TMS_ACCT_STA_LST_UPDT", "2020-06-17 15:06:01.29");
        dataPoint.put("CDE_ACCT_TYP", "ttttt3");
        dataPoint.put("NUM_ACCT", "1234567");
        dataPoint.put("CDE_FIN_INST", "ttttt5");
        dataPoint.put("afDPFireTS", "2020-06-17 15:06:01.29");
        dataPoint.put("IDN_DLET_USER", "ttttttttttttttttttt9");
        dataPoint.put("IDN_ACCT", "ttttttttttttttttttttttttttttttttttttttttttttttttt6");
        dataPoint.put("CDE_ACCT_STA", "Y");
        dataPoint.put("CDE_ACCT_SUB_TYP", "ttttt2");
        dataPoint.put("IND_DLET", "1");
        dataPoint.put("IDN_LST_UPDT_USER", "tttttttttttttttttt10");
        dataPoint.put("NUM_FA", "111");
        dataPoint.put("TMS_XTRNL_ACCT_FEFR_ID", "2020-06-17 15:06:01.29");
        dataPoint.put("NUM_OFFICE", "222");
        dataPoint.put("CDE_ACCT_ENTITY", "tttttttttttttt1");
        dataPoint.put("IDN_CREAT_PGM", "ttttttttt7");
        dataPoint.put("EventProcessedUtcTime", "2020-06-17T17:21:35.5692228Z");
        dataPoint.put("PartitionId", "10");
        dataPoint.put("EventEnqueuedUtcTime", "2020-06-17T15:06:01.3070000Z");

        for (int i = 0; i < 4; i++) {

            String CDE_ACCT_ENTITY = (String) dataPoint.get("CDE_ACCT_ENTITY");
            String CDE_ACCT_STA = (String) dataPoint.get("CDE_ACCT_STA");
            String CDE_ACCT_SUB_TYP = (String) dataPoint.get("CDE_ACCT_SUB_TYP");
            String CDE_ACCT_TYP = (String) dataPoint.get("CDE_ACCT_TYP");
            String CDE_DATA_SRC = (String) dataPoint.get("CDE_DATA_SRC");
            String CDE_FIN_INST = (String) dataPoint.get("CDE_FIN_INST");
            String IDN_ACCT = (String) dataPoint.get("IDN_ACCT");
            int IDN_ACCT_ROW = Integer.parseInt((String) dataPoint.get("IDN_ACCT_ROW"));
            String IDN_CREAT_PGM = (String) dataPoint.get("IDN_CREAT_PGM");
            String IDN_CREAT_USER = (String) dataPoint.get("IDN_CREAT_USER");
            String IDN_DLET_USER = (String) dataPoint.get("IDN_DLET_USER");
            String IDN_LST_UPDT_USER = (String) dataPoint.get("IDN_LST_UPDT_USER");
            String IDN_LST_UPDT_PGM = (String) dataPoint.get("IDN_LST_UPDT_PGM");
            String IND_DLET = (String) dataPoint.get("IND_DLET");
            BigDecimal NUM_ACCT = new BigDecimal((String) dataPoint.get("NUM_ACCT"));
            NUM_ACCT = NUM_ACCT.setScale(7);
            BigDecimal NUM_FA = new BigDecimal((String) dataPoint.get("NUM_FA"));
            NUM_FA = NUM_FA.setScale(3);
            BigDecimal NUM_OFFICE = new BigDecimal((String) dataPoint.get("NUM_OFFICE"));
            NUM_OFFICE = NUM_OFFICE.setScale(3);
            Timestamp TMS_ACCT_STA_LST_UPDT = Timestamp.valueOf((String) dataPoint.get("TMS_ACCT_STA_LST_UPDT"));
            Timestamp TMS_CREAT = Timestamp.valueOf((String) dataPoint.get("TMS_CREAT"));
            Timestamp TMS_DLET = Timestamp.valueOf((String) dataPoint.get("TMS_DLET"));
            Timestamp TMS_LST_UPDT = Timestamp.valueOf((String) dataPoint.get("TMS_LST_UPDT"));
            Timestamp TMS_XTRNL_ACCT_FEFR_ID = Timestamp.valueOf((String) dataPoint.get("TMS_XTRNL_ACCT_FEFR_ID"));
            Timestamp afDPFireTS = Timestamp.valueOf((String) dataPoint.get("afDPFireTS"));

            statement.setString(1, CDE_ACCT_ENTITY);
            statement.setString(2, CDE_ACCT_STA);
            statement.setString(3, CDE_ACCT_SUB_TYP);
            statement.setString(4, CDE_ACCT_TYP);
            statement.setString(5, CDE_DATA_SRC);
            statement.setString(6, CDE_FIN_INST);
            statement.setString(7, IDN_ACCT);
            statement.setInt(8, IDN_ACCT_ROW);
            statement.setString(9, IDN_CREAT_PGM);
            statement.setString(10, IDN_CREAT_USER);
            statement.setString(11, IDN_DLET_USER);
            statement.setString(12, IDN_LST_UPDT_USER);
            statement.setString(13, IDN_LST_UPDT_PGM);
            statement.setString(14, IND_DLET);
            statement.setBigDecimal(15, NUM_ACCT, 7, 0);
            statement.setBigDecimal(16, NUM_FA, 3, 0);
            statement.setBigDecimal(17, NUM_OFFICE, 3, 0);
            statement.setTimestamp(18, TMS_ACCT_STA_LST_UPDT);
            statement.setTimestamp(19, TMS_CREAT);
            statement.setTimestamp(20, TMS_DLET);
            statement.setTimestamp(21, TMS_LST_UPDT);
            statement.setTimestamp(22, TMS_XTRNL_ACCT_FEFR_ID);

            statement.setTimestamp(23, afDPFireTS);
            statement.setTimestamp(24, afDCFireTS);
            statement.setTimestamp(25, null);

            statement.addBatch();

        }
        statement.executeBatch();

        connection.commit();
        statement.clearBatch();
        System.out
                .println("batch insert for data successfull for batch with execution time of " + afDCFireTS.toString());
    }
}
```

**Comment by ulvii** [comment]
If you want us to further investigate this, please provide full vanilla JDBC code so that I can reproduce  the issue.

**Comment by shayannyc25** [comment]
i am not sure what you mean by 'full vanila jdbc code'. i had given you the Azure Java Function code. As i had suspected, you couldn't reproduce this in single threaded java app. I strongly believe this issue is threading related, as azure function run time environment creates multiple instances of the java azure function class and calls the method in each instance in parallel. Unless you also create a java azure function, you may not be able to reproduce this issue. i am attaching the whole java file that has the azure java function.  You need to create an Event hub hub with 20 partitions and then just use the sample data file i had given to put events (at least 1000 ,if possible 10,000) in the event hub and then create a java azure function
[EventHubTriggerJavaToSQLDB_MS_Debug.txt](https://github.com/microsoft/mssql-jdbc/files/4804818/EventHubTriggerJavaToSQLDB_MS_Debug.txt)

 using the code i have attached and run the function on your desktop using the Visual Studio Code

**Comment by ulvii** [comment]
I will try to setup an environment based on your description sometime next week, but if you can create a reproduction that doesn't involve Azure functions and is standalone JDBC code, it would speed up the investigation process a lot.

**Comment by peterbae** [comment]
Hi @shayannyc25, just letting you know that I was able to reproduce the issue on Azure Java Function. We'll let you know once we have an update on what's causing this issue.

**Comment by shayannyc25** [comment]
hi @peterbae  , glad that you were able to reproduce it. Appreciate your help.

**Comment by peterbae** [comment]
Hi @shayannyc25, I've made a fix for this in PR #1378. My local testing showed that it fixed your issue as well as issue #1301. Could you try with these jars and see if it fixes the problem for you? Thanks.
[github-1301.zip](https://github.com/microsoft/mssql-jdbc/files/4867208/github-1301.zip)

**Comment by peterbae** [comment]
Oops, it looks like GitHub automatically closed this issue when we merged PR #1378, but we haven't received confirmations regarding this issue. @shayannyc25 were you able to test if this fixes your issues? Thanks.

**Comment by shayannyc25** [comment]
@peterbae  sorry was out for few days and just got back today. I was able to test and was able to verify the fix (for SQLServerPreparedStatement.executeBatch()) in my Java Azure Function. I tested it with the latest Azure Functions run time update that became available few days ago.  Ticket can be closed. Thanks. When will new jbdc driver for java 8 and 11 with this fix be made available in  https://mvnrepository.com ?
thanks

**Comment by peterbae** [comment]
Thanks for confirming. We're aiming to deploy the next fix (v8.4.0) by the end of this month.
